### PR TITLE
enhancement(socket sink): Add IPv6 supports

### LINKED ISSUE #3573: Socket sinks fail to parse IPv6 address `[host]:<port>`
`http::Uri::host` returns `[host]` which creates an error when passing to `ToSocketAddrs`.

```toml
[sinks.out]
mode = "udp"
type = "socket"
address = "[::1]:5000"
inputs = ["in"]
encoding.codec = "json"
```

Solution: trim left and right brackets.

### LINKED ISSUE #3574: UDP socket sink does not work for IPv6
As documented in [send_to](https://doc.rust-lang.org/std/net/struct.UdpSocket.html#method.send_to),

> This will return an error when the IP version of the local socket does not match that returned from ToSocketAddrs.

UDP sinks bind to IPv4 by default, therefore sending to an IPv6 will fail.

Reproducible:

```toml
[sinks.out]
mode = "udp"
type = "socket"
address = "test.host:5000"
inputs = ["in"]
encoding.codec = "json"
```

/etc/hosts:
```text
::1 test.host
```

Output:
```
ERROR connection{host=test.host port=5000}: vector::sinks::util::udp: send failed error=Address family not supported by protocol (os error 97)
```

### PR DESCRIPTION
**Pull request by juchiast**
fixes #3573 
fixes #3574 

Signed-off-by: Duy Do <juchiast@gmail.com>

### CODE REVIEW

**Comment by binarylogic** [comment]
Can we update the `endpoint` docs in the relevant `.meta/*.toml` files to mention that IPv4 and IPv6 is supported?

**Comment by juchiast** [comment]
@binarylogic I don't think it is necessary, as a user, I always assume that both versions are supported.

File: src/dns.rs [hunk 1]
```diff
@@ -37,11 +37,18 @@ impl Resolver {
...
-                .await
-                .context(JoinError)?
-                .map(LookupIp)
-                .context(UnableLookup)
+            spawn_blocking(move || {
+                let mut name_ref = name.as_str();
+                // strip IPv6 prefix and suffix
+                if name_ref.starts_with('[') && name_ref.ends_with(']') {
+                    name_ref = &name_ref[1..name_ref.len() - 1];
+                }
```
[the remark this answers is not in the document]
Comment (juchiast) [reply src/dns.rs]: Thanks for the suggestion!

File: src/sinks/util/udp.rs [hunk 2]
```diff
@@ -137,7 +117,21 @@ impl UdpSink {
                         State::Backoff(self.next_delay01())
                     }
                 },
-                State::ResolvedDns(addr) => return Ok(Async::Ready(addr)),
+                State::ResolvedDns(addr) => {
+                    let bind_address = find_bind_address(&addr);
+                    let socket = match UdpSocket::bind(bind_address) {
+                        Ok(socket) => socket,
+                        Err(_) => unreachable!(),
```
[the remark this answers is not in the document]
Comment (juchiast) [reply src/sinks/util/udp.rs]: I must have used this as an `unimplemented!` and then forgot about it. Should properly use `unimplemented!` next time!

**Comment by juchiast** [comment]
`defers_at_high_concurrency` failed.

**Comment by daddyzilla33** [comment]
Yes

**Review by lukesteensen — APPROVED** [review]

**Review by JeanMertz — APPROVED** [review]

**Pull request merged**